Support for secret parameters in Instance

nilebox opened the pull request:
See issue [#621](https://github.com/kubernetes-incubator/service-catalog/issues/621) for details.
Implementation described in https://github.com/kubernetes-incubator/service-catalog/issues/621#issuecomment-305356537:

```go
// InstanceSpec represents the desired state of an Instance.
type InstanceSpec struct {
	...

	// Optional: AlphaSecretParameters is reference to the secret object containing
	// sensitive information to pass to the underlying broker. This may be
	// empty if no secret object is specified.
	// +optional
	AlphaSecretParameters *v1.LocalObjectReference `json:"alphaSecretParameters,omitempty"`
}
```
- Only one of `Parameters` and `AlphaSecretParameters` can be specified at a time (no merging strategy, no separate field in OSB Instance request)
- Added unit and integration tests for `AlphaSecretParameters` based on the ones for `Parameters`

(substitutes https://github.com/kubernetes-incubator/service-catalog/pull/957)

nilebox commented:
/cc @pmorie @duglin this is a replacement for https://github.com/kubernetes-incubator/service-catalog/pull/957 with no code changes (Paul suggest to open a new PR to get rid of all the obsolete comments)

pmorie commented on pkg/apis/servicecatalog/checksum/unversioned/checksum.go:
```diff
@@ -38,6 +39,10 @@ func InstanceSpecChecksum(spec servicecatalog.InstanceSpec) string {
 		specString += fmt.Sprintf("parameters:\n\n%v\n\n", string(spec.Parameters.Raw))
 	}
+	if spec.AlphaSecretParameters != nil {
+		// TODO: calc checksum of secret contents instead?
```
I think technically you have to calculate both the checksum based on the spec and the checksum based on the values of the parameters.  The tricky part is that you have to calculate the checksum of the parameters based on the value of the parameters you set.  I think that implies a separate checksum field that the controller has to set that is based on the parameter values.

Note that if any of these values change due to user update, there's a bug in the controller because we do not implement update :) Someone should go and do that.

nilebox replied:
> if any of these values change due to user update, there's a bug in the controller because we do not implement update

Exactly. Thus I would prefer the secret contents checksum to be addressed along with secret update support in a separate PR (once we find the brave contributor willing to do this). TODO should be sufficient for now to remind and embarass us :)

arschles replied:
>I would prefer the secret contents checksum to be addressed along with secret update support in a separate PR

I agree. I have missed the last few design meetings, but from what I understand, the group is planning to address secret updates in a few phases. This area should fall into that work.

That said, @nilebox can you add an issue for this specific work item, or perhaps mention it in an existing secret if appropriate?

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -266,10 +266,16 @@ type InstanceSpec struct {
 	// passed to the underlying broker.
 	Parameters *runtime.RawExtension
 
+	// Optional: AlphaSecretParameters is reference to the secret object containing
```
Can you explain in the godoc for `Parameters` and `AlphaSecretParameters` that only one of these fields can be set at a time?  Also, explain the behavior how the secret is translated into the parameter payload.

duglin replied:
Is this true?

arschles replied:
@duglin if both can be set at the same time, we'll have to define rules for merging. And I would like us to do that as part of a design discussion/proposal process, and keep out of scope for this PR.

pmorie commented on pkg/controller/controller.go:
```diff
@@ -511,6 +511,26 @@ func unmarshalParameters(in []byte) (map[string]interface{}, error) {
 	return parameters, nil
 }
 
+func fetchSecretParameters(kubeClient kubernetes.Interface, namespace string, name string) (map[string]interface{}, error) {
```
godoc for this method?

pmorie commented on pkg/controller/controller_instance.go:
```diff
@@ -251,7 +251,20 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) error {
-	if instance.Spec.Parameters != nil {
+	if instance.Spec.Parameters != nil && instance.Spec.AlphaSecretParameters != nil {
+		// We don't support merging of plain and secret parameters, thus only one of them could be set
+		s := "At least one of 'Parameters' and 'AlphaSecretParameters' fields must be null"
```
There should be an API validation as well.

pmorie commented on test/integration/clientset_test.go:
```diff
@@ -466,6 +466,9 @@ func testInstanceClient(sType server.StorageType, client servicecatalogclient.In
 			ServiceClassName: "service-class-name",
 			PlanName:         "plan-name",
 			Parameters:       &runtime.RawExtension{Raw: []byte(instanceParameter)},
+			AlphaSecretParameters: &v1.LocalObjectReference{
```
This shouldn't be a valid Instance; see other comments about adding a validation.

kibbles-n-bytes commented:
/retest

nilebox commented:
@pmorie see the commit https://github.com/kubernetes-incubator/service-catalog/pull/1028/commits/4e10c5d060c0bd1bcf4dafceff4942433a7eb96a addressing your comments

spadgett commented on pkg/controller/controller_test.go:
```diff
@@ -284,6 +286,25 @@ const instanceParameterSchemaBytes = `{
 }`
 
 // broker used in most of the tests that need a broker
```
Looks like this comment is for a different function?

nilebox replied:
@spadgett good catch, will fix.

duglin commented on pkg/apis/servicecatalog/validation/instance.go:
```diff
@@ -60,6 +60,10 @@ func validateInstanceSpec(spec *sc.InstanceSpec, fldPath *field.Path, create boo
 	}
 
+	if spec.Parameters != nil && spec.AlphaSecretParameters != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("alphaSecretParameters"), spec.AlphaSecretParameters, "only one of 'Parameters' and 'AlphaSecretParameters' can be set at a time"))
```
I seem to recall that there was concern in the past about secrets being limited in size so would forcing everything into the secret when perhaps just one field might need it be a concern?  Additionally, why don't we want to allow both?  Is this due to the conflict's name concern?

arschles replied:
I also mentioned this above, but I'll repeat here for completeness. There are space constraints for secrets and I think it would be appropriate to design merge rules for doing both. But I also think it's appropriate to do so outside of the scope of this PR and submit a follow-up to this PR once the design process has finished for that work. Thoughts?

nilebox replied:
@duglin @arschles Secret contains a key-value map here (each will be converted into top-level field in the payload), with every value stored separately I believe. I don't think that multiple secrets will gain anything regarding the size limit there.
This is exactly the same approach as the one used for Bindings already (just in reverse direction), so I wouldn't be concerned about the size limit there. If this approach turns out to have size limitations, it shall be fixed for both Bindings and Instances.

duglin replied:
the suggestion I heard in the past wasn't for multiple secrets, it was for some params to be a non-secret config map.

nilebox replied:
@duglin @arschles re: merging
I am not against support of both secret and non-secret parameters simultaneously in general.
But I would prefer to keep this out of scope of this PR. This PR is about design for secret parameters support, and lack of merging strategy makes it easier to review.
A separate PR with adding merging support is fine with me, and will also be focused on single feature easy to review, and won't require any breaking API changes.

nilebox replied:
> the suggestion I heard in the past wasn't for multiple secrets, it was for some params to be a non-secret config map.

Makes sense. Then it should be addressed by a separate PR adding merging strategy support between secret and non-secret parameters.

arschles replied:
@nilebox would you care to write up a proposal for merging? If not, that's fine and I will take that on.

duglin commented:
overall it seems good to me, just one question about normal params vs secret one.

arschles commented on pkg/controller/controller.go:
```diff
@@ -512,6 +512,27 @@ func unmarshalParameters(in []byte) (map[string]interface{}, error) {
+	for k, v := range secret.Data {
+		var parameters interface{}
+		err := json.Unmarshal(v, &parameters)
+		if err != nil {
```
nit: can you please combine this and the above line into one?

arschles replied:
I won't block this PR on this stylistic point :)

arschles commented on pkg/controller/controller_instance_test.go:
```diff
@@ -336,6 +339,90 @@ func TestReconcileInstanceWithInvalidParameters(t *testing.T) {
+	assertInstanceReadyTrue(t, updatedInstance)
+
+	updateObject, ok := updatedInstance.(*v1alpha1.Instance)
+	if !ok {
```
thanks for doing this check :+1:

spadgett commented:
The secret key should match the parameter property name, correct?

Secret keys (regex `[-._a-zA-Z0-9]+`) are more restrictive than JSON object property names, which can have any character. So parameters that have other characters in the property name can't be added to a secret.

I ran into this prototyping against a real broker that uses a `/` in a parameter name.

cc @bparees

nilebox commented:
@spadgett interesting, wasn't aware of that limitation.
However, as I said before, the design of the instance secret parameters support directly follows the approach that was previously taken for storing credentials produced by the binding.
See the code [controller_binding.go#L399](https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_binding.go#L399)
The credentials object is an arbitrary JSON as well (see the [OSB spec](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#body-7)). So if we decide to change the design for instance parameters, I would sync it with updating the binding credentials design as well.

> I ran into this prototyping against a real broker that uses a / in a parameter name.

wow, that is concerning. Have never seen `/` in the JSON property names before.
@duglin @arschles @pmorie what do you think?
[Google JSON guidelines](https://google.github.io/styleguide/jsoncstyleguide.xml#Property_Name_Format) don't allow `/`, for example:
> The first character must be a letter, an underscore (_) or a dollar sign ($).
Subsequent characters can be a letter, a digit, an underscore, or a dollar sign.

The only alternative solution I see there is to put the entire payload into a single Secret **key** (as the secret _value_ is an arbitrary raw byte array).

spadgett commented:
Yeah, I don't think this is common, but I thought I'd raise the issue since I ran into it. It looks like the Google guidelines you link allow properties that start with `$`, which would also be a problem here.

duglin commented:
Is there any reason NOT to store the entire parameters JSON as a blob? In this case since the intended user of this data isn't kube, its something outside of kube, and that something may require non-kube-acceptable (but valid json) key names it seems like storing it as a blob is the easiest. Any kind of mapping will just be annoying because then we need to deal with things like escaping, etc...  Besides, they gave us a json blob, when they GET the resource they'll be given a json blob - seems right.

nilebox commented:
> Is there any reason NOT to store the entire parameters JSON as a blob?

@duglin I personally don't have any objection to storing the entire JSON as a single blob given the limitations of current proposal mentioned above. Basically just will need to replace the `v1.LocalObjectReference` with `v1.SecretKeySelector`.

The only my concern here is that it means you won't be able to directly inject the output credentials from Bindings as an input for Instance (you'll have to merge multiple secret key-value pairs into a single blob first). For us it's fine, not sure how it works for @kibbles-n-bytes and @vaikas-google ?

duglin commented:
hmmm, perhaps the vcap_services blob isn't looking so bad after all.  :-)

sometimes I do wonder if we should store the creds from the broker as a blob AND as individual properties (when possible) since not all of them could be stored in our maps anyway. You could then use the blob as input into the next service in your chain.

nilebox commented:
> as a blob AND as individual properties

@duglin sounds tempting, but my gut tells me that I need to resist this temptation 😀 
Having duplicate data could lead to the other issues like atomicity I guess.
Can definitely discuss this option tomorrow.

duglin commented:
Well, to be honest, I was thinking that the blob was needed due to the key-name restriction we're talking about. I'm not thrilled with the idea of either dropping some key or defining some name mapping. This means that the only way to guarantee a user doesn't lose data is to see the original JSON as a blob.  If we had to choose just one of the two forms... I'd pick the blob.

In previous discussions we talked about doing things with well-defined cred properties (like using them to define Kube services), and so it might make sense to just pull out that subset of properties and only create those - we just need to make sure that the names defined in the OSB API spec are also Kube-approved.

nilebox commented:
@duglin alright, let me go offtopic and crazy a bit too (you started this!).
What if the Binding produces a single raw blob secret "as is", without any implications of later using separate credential keys as environment variables for the Pods (as we are moving PodPreset out of Binding). Then interpretation of this blob is a concern of the separate resource wrapping it.

Then PodPresetBinding (specifically designed for injecting credentials as environment variables) would take this blob secret as an input and produce a key-value map with individual properties, as long as make sure that keys match environment variables format (uppercase, only latin, digits and underscode etc) and rename them to match as necessary (or even explicitly renaming them as part of PodPresetBinding spec). We could also define a set of the only keys we need to extract from the blob (and ignoring the rest).

Thoughts?

duglin commented:
I don't think that's not much different than what I was suggesting except you're doing the mapping/tweaking/subsetting during PodPresetBinding time instead of during the "secret creation time", right? I guess the advantage of that is that other PPB type of things might have different mapping rules - but is that good?  Or should we push for consistency?  Not sure its good that PPB and fooBinding expose a different set of properties simply due to charset-validity issues/mappings.

If we want to avoid forcing apps to deal with raw json as much as possible then it seems logical that we should allow people to access those properties directly via the secret instead of forcing them to use a PPB type of thing.  But is avoiding json parsing a requirement?  I don't know yet - I mean CF has gotten by w/o it up to now.  Again.... vcap-services isn't looking so bad any more!  LOL  it really does solve many issues, including name conflicts.

One thing we can consider, not for now but for the future, is to change the spec to force people to limit the charset of the properties to valid env var values since exposing these as env var seems like its going to be a fairly common usage pattern.  We can then also define (in the spec) a *recommendation* for how to deal with name conflicts - e.g. instanceName_property.

pmorie commented:
> Is there any reason NOT to store the entire parameters JSON as a blob?

I think this is valid mode, but this PR is not at all related to that.  I've opened this issue to discuss, please move discussion there: https://github.com/kubernetes-incubator/service-catalog/issues/1045

pmorie commented:
Adding do-not-merge label since we may change this PR slightly based on design discussions in the SIG

nilebox commented:
@pmorie @arschles @spadgett @kibbles-n-bytes @duglin @ash2k
Please see the implementation of revised proposal (with comment https://github.com/kubernetes-incubator/service-catalog/issues/1048#issuecomment-316557027). Does it look good for you?

duglin commented:
I think the CI failure is real:
```
The Instance "ups-instance" is invalid: Spec.parameters: Invalid value: servicecatalog.Parameters{SecretRef:(*v1.LocalObjectReference)(nil), SecretKeyRef:(*v1.SecretKeySelector)(nil)}: Parameters must not be empty if defined
/var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/contrib/hack/test_walkthrough.sh: line 249: Error when creating ups-instance.
```

duglin commented:
removing do-not-merge since I think we've resolved any lingering design issues

duglin commented on pkg/apis/servicecatalog/checksum/unversioned/checksum.go:
```diff
@@ -37,7 +37,12 @@ func InstanceSpecChecksum(spec servicecatalog.InstanceSpec) string {
-		specString += fmt.Sprintf("parameters:\n\n%v\n\n", string(spec.Parameters.Raw))
+		params := spec.Parameters
+		if params.SecretRef != nil {
+			specString += fmt.Sprintf("parameters.secretRef: %v\n", params.SecretRef.Name)
```
Am I correct that this means that if we're recalculating the checksum (for whatever reason) we won't detect a change in the secret's data?  Is this the normal behavior for kube?  Feels a little odd that we won't detect a change in the secret values yet if something else causes us to use the secert in a request we will grab the new values. Almost like we're 1/2 blind to the new data.  :-)

duglin replied:
If this really is ok, can we add a comment to that effect here so the next person looking at the code doesn't think its a design flaw, rather its done on purpose?

nilebox replied:
@duglin that's right, see the comment: https://github.com/kubernetes-incubator/service-catalog/pull/1028#discussion_r127391333
I can add a comment with a link to a separate issue there.

duglin replied:
thanks

duglin commented on pkg/apis/servicecatalog/checksum/versioned/v1alpha1/checksum.go:
```diff
@@ -37,7 +37,12 @@ func InstanceSpecChecksum(spec v1alpha1.InstanceSpec) string {
-		specString += fmt.Sprintf("parameters:\n\n%v\n\n", string(spec.Parameters.Raw))
+		params := spec.Parameters
+		if params.SecretRef != nil {
+			specString += fmt.Sprintf("parameters.secretRef: %v\n", params.SecretRef.Name)
```
Same comment/request here.

duglin commented:
I now our docs are probably way behind the code, but is there a good spot in our docs where we could add info on this so that people can read-up on how to use secrets for parameters w/o having to read the code?

nilebox commented:
@duglin do we have any docs about instance and binding creation apart from the walkthrough at all? If there is, I haven't seen one :)

duglin commented:
we have design.md - that I'm sure is really out of date, but better than nothing.

duglin commented on pkg/apis/servicecatalog/testing/fuzzer.go:
```diff
@@ -65,6 +66,17 @@ func createParameter(c fuzz.Continue) (*runtime.RawExtension, error) {
 	return &runtime.RawExtension{Raw: b}, nil
 }
 
+func createParameters(c fuzz.Continue) *servicecatalog.Parameters {
```
Since this seems to only be used for instances, and you renamed the other one createBindingParameters, how about calling this one createInstanceParameters just for consistency/clarity?  minor nit.

nilebox replied:
@duglin Yeah though about that as well, but then deliberately chose this name to reuse this method for bindings as well in the future (as a separate PR of course).
Still agree that probably better to call it `createInstanceParameters` for now and rename after if needed.

duglin commented on pkg/controller/controller.go:
```diff
@@ -512,6 +514,42 @@ func unmarshalParameters(in []byte) (map[string]interface{}, error) {
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]interface{})
```
Sorry for piggy-backing this question on this PR, but I have a golang question... is there a fundamental difference between:
```
result := make(map[string]interface{})
```
and
```
result := map[string]interface{}{}
```
I don't think there is but I just wanted to check.

nilebox replied:
@duglin yep, it's the same (since we don't explicitly pass extra capacity in the `make` call).
I think it's more about the convention and patterns of usage, see https://groups.google.com/forum/#!topic/golang-nuts/YnqSi2yURSY
```
Similar to new(Struct) vs Struct{}, I use them in slightly different ways:
1) If the value is done and I don't need to add anything to it before releasing it off into the world, I use the {} version
2) If I'm planning on adding things to it (usually programmatically) I use the make() (or new()) version.
```
Since I plan to fill this map below, I'm using `make` there.

duglin replied:
thanks

ash2k replied:
You could pre-allocate `len(secret.Data)` elements

duglin commented on pkg/controller/controller.go:
```diff
@@ -512,6 +514,42 @@ func unmarshalParameters(in []byte) (map[string]interface{}, error) {
+	result := make(map[string]interface{})
+	for k, v := range secret.Data {
+		var parameters interface{}
+		if json.Unmarshal(v, &parameters) != nil {
```
isn't the data a base64 encoded string, not json?

nilebox replied:
@duglin no, it's raw byte array there. If I recall correctly, the transformation into/from base64 is being done by `kubectl` on-the-fly (when you ask for `-o yaml`), and you won't see it with raw REST API.

See binging credentials code, for example - it doesn't use base64 encode either: https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/controller/controller_binding.go#L402

nilebox replied:
>  and you won't see it with raw REST API.

actually you probably will, since there is no way of passing raw byte array in JSON. still, I don't think we need to deal with base64 when we use client-go for working with secrets.

duglin replied:
I was looking at the client code (hopefully the right spot), and saw this:
```
    // Data contains the secret data. Each key must be a valid DNS_SUBDOMAIN
    // or leading dot followed by valid DNS_SUBDOMAIN.
    // The serialized form of the secret data is a base64 encoded string,
    // representing the arbitrary (possibly non-string) data value here.
    // Described in https://tools.ietf.org/html/rfc4648#section-4
    // +optional
    Data map[string][]byte `json:"data,omitempty" protobuf:"bytes,2,rep,name=data"`
```
ok, if its not base64 encoded, does the data have to be json and not just a normal string?

nilebox replied:
@duglin as long as the string can be translated into `map[string]interface{}` we don't care about the string format. if we want to explicitly require it to be JSON, I would avoid any unmarshalling completely, and just pass the raw bytes then (will need to change the data types for parameters in the instance request though).

duglin replied:
I probably wasn't clear. My concern is that you call json.Unmarshal() on the secret key's value, which means if its not json we'll error out. Shouldn't we just pass the value "as is" w/o trying to convert it from json?

nilebox replied:
if it's a JSON, I don't see how json.Unmarshal() to `map[string]interface{}` can fail. It just transforms JSON into generic map of maps (and so on). See [unstructured.go](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/unstructured/unstructured.go#L52) for example, it utilizes `map[string]interface{}` for working with any Kubernetes resource as arbitrary JSON without knowing its type.

The reason for unmarshalling is just a convenient way of representing JSON object in Go from any source (Secret with multiple keys, single Secret key, and possibly more in the future).

nilebox replied:
> it could be a primitive value (int, string etc)

actually we can't support this with multiple Secret keys (as a flat map).
I have pointed to this limitation many times before, see https://github.com/kubernetes-incubator/service-catalog/issues/621#issuecomment-302833747.
You can't just say "I dunno, maybe it's JSON, or maybe it's a string", you have to choose one upfront in the code. We decided to choose JSON, and for types like `map[string]string` we should probably use single Secret key with JSON blob.

nilebox replied:
we can ultimately decide to treat a Secret with multiple keys as `map[string]string` though, and ditch support for JSON subtrees there. For complex JSON then there will be only option of having a single secret key blob.
@duglin what do you think?

nilebox replied:
I want to discuss this at the call tomorrow

duglin replied:
I may have confused myself so sorry if this doesn't make sense, but the secret that we're pointing to from Instances can be user-defined, right?  Meaning, an end-user can create a secret (basically a map) and then references that from the Instance.parameters field. The user could put any data they want as the value of each entry in that map - it doesn't need to be json. So, wouldn't you trying to deserialize as json cause a problem?

duglin commented on pkg/controller/controller_instance.go:
```diff
@@ -252,16 +252,46 @@ func (c *controller) reconcileInstance(instance *v1alpha1.Instance) error {
+					v1alpha1.InstanceConditionReady,
+					v1alpha1.ConditionFalse,
+					errorWithParameters,
+					"Error fetching instance secret parameters. "+s,
```
Minor thing but shouldn't we just use 's' here instead of repeating that we couldn't fetch the secret? This will appear as: "Error fetching instance secret parameters. Failed to fetch Instance secret parameters..."

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -439,3 +439,15 @@ const (
+// - Instance input parameters
+// - Binding output credentials
+type Parameters struct {
+	// Reference to a secret in the namespace.
```
This removes the ability to do inline parameters -- which I suspect might break some users.  Should we include the ability to use raw (maybe call it 'embedded' or something like that) with a note that secret parameters should come from a secret?

nilebox replied:
ok, will add support for `Raw` parameters.

nilebox replied:
@pmorie do you think it should be represented as YAML or JSON? Just found that current implementation requires YAML...

nilebox replied:
Added to the agenda to discuss tomorrow.

pmorie replied:
I think keeping parity with what we currently have is fine.  There is not a way, afaik, to only support json payloads in these fields.

pmorie commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -439,3 +439,15 @@ const (
+	SecretRef *v1.LocalObjectReference
+	// Reference to a secret key in the namespace.
+	// The parameters will be set from the JSON contained in this secret key
+	SecretKeyRef *v1.SecretKeySelector
```
I actually think that we should use a distinct type here, because the `SecretKeySelector` has the `Optional` field whose semantics we don't want.

nilebox replied:
Yeah the `Optional` field concerned me as well. Okay, will replace with our own type then.

pmorie commented on pkg/controller/controller_instance_test.go:
```diff
@@ -215,7 +214,70 @@ func TestReconcileInstanceNonExistentServicePlan(t *testing.T) {
+	// verify no kube resources created
+	assertActionsVerb(t, kubeActions, "get")
+	// only 1 action - get secret
+	assertNumberOfActions(t, kubeActions, 1)
```
You can use this construction to test the actions:

```go
	if err := checkKubeClientActions(kubeActions, []kubeClientAction{
		{verb: "get", resourceName: "namespaces", checkType: checkGetActionType},
	}); err != nil {
		t.Fatal(err)
	}
```

pmorie commented on pkg/controller/controller_instance_test.go:
```diff
@@ -264,10 +345,12 @@ func TestReconcileInstanceWithParameters(t *testing.T) {
+	assertActionsVerb(t, kubeActions, "get")
+	// one action comes from getting namespace uid
+	// another action comes from getting secret
+	assertNumberOfActions(t, kubeActions, 2)
```
Test the actions per above comment.

pmorie commented on pkg/controller/controller_test.go:
```diff
@@ -1107,6 +1148,31 @@ func testNumberOfActions(t *testing.T, name string, f failfFunc, actions []clien
+	testActionsVerb(t, "" /* name */, fatalf, actions, verb)
+}
+
+func expectActionsVerb(t *testing.T, name string, actions []clientgotesting.Action, verb string) bool {
```
Seems like there's some overlap with `checkKubeActions` here.

nilebox replied:
I still see the value of such generic method for making sure that only `get` requests were performed (no side-effects), for example.

nilebox replied:
Ok, deleted those methods to avoid duplicate checks :)

ash2k commented on pkg/apis/servicecatalog/types.go:
```diff
@@ -449,5 +449,13 @@ type Parameters struct {
+// SecretKeyReference references a key of a Secret.
+type SecretKeyReference struct {
+	// The name of the secret in the pod's namespace to select from.
+	v1.LocalObjectReference
```
Don't do this - `SecretKeyReference` will inherit all the methods from `LocalObjectReference` and you absolutely do not want that.

ash2k commented on pkg/apis/servicecatalog/v1alpha1/types.go:
```diff
@@ -450,5 +450,13 @@ type Parameters struct {
+// SecretKeyReference references a key of a Secret.
+type SecretKeyReference struct {
+	// The name of the secret in the pod's namespace to select from.
+	v1.LocalObjectReference `json:",inline"`
```
Same as in another file.

ash2k commented on pkg/controller/controller.go:
```diff
@@ -512,6 +514,42 @@ func unmarshalParameters(in []byte) (map[string]interface{}, error) {
 
+// fetchSecretParameters requests and returns the contents of the given secret as a map
+func fetchSecretParameters(kubeClient kubernetes.Interface, namespace string, secretRef *v1.LocalObjectReference) (map[string]interface{}, error) {
+	if kubeClient == nil {
```
Just curious - why is this necessary?

duglin commented on contrib/examples/walkthrough/ups-instance.yaml:
```diff
@@ -7,6 +7,7 @@ spec:
+    inline:
+      credentials:
+        name: root
+        password: letmein
```
Having the "inline" property in there feels a little clunky. Is it not possible to do something like:
```
spec:
  serviceClassName: user-provided-service
  planName: default
  parameters:
    credentials:
      name: root
      password: letmein
    k8s_secretRef:
      name: mySecret
    k8s_secretKeyRef:
      name: myOtherSecret
      key: key1
```
Basically just have some well-defined property names that we reserve. Note, that if we do this there's no reason not to allow for all 3 variants to appear at the same time.

Also, it seems a little odd that we've said that there may be sensitive data in the binding credentials along side non-sensitive data (so we're asking for there to be a way to identify each in the spec), and yet we're not providing a way for someone to do the same thing on the input side of things - if we adopt a "union only" approach.  We can deal with any potential conflicts by just making it clear what the rules are - e.g. error out on conflict.

duglin replied:
Also, any reason we don't allow the creds to be stored in a configMap?  Guess that could be done post-beta as long as we provide the right extension point.

nilebox replied:
> there's no reason not to allow for all 3 variants to appear at the same time

The reason is that supporting this is harder, and we have to come up with merging rules convention. I am not against it, but I would like to make it a separate PR once we come to an agreement. It won't be a breaking change, just a more flexible extension on top of this one.

> Also, any reason we don't allow the creds to be stored in a configMap?

No reason, just want to keep this PR simple just for the pattern of supporting different ways of representing parameters. For configMap (and any other modes) we can raise separate PRs.

nilebox replied:
@duglin just for the clarity: you surely can add support for multiple sources of data for parameters (but it's a separate issue to be designed and addressed in a separate PR):
```
spec:
  serviceClassName: user-provided-service
  planName: default
  parameters:
    instance:
      credentials:
        name: root
        password: letmein
    secretRef:
      name: mySecret
    secretKeyRef:
      name: myOtherSecret
      key: key1
```
I am strongly against any special prefixes like `k8s_*`
The merging strategy is a separate issue, but I would support merging only at the top level.

nilebox replied:
@duglin My alternative proposal to address your concern:
```
spec:
  serviceClassName: user-provided-service
  planName: default
  parameters:
    username: root
    password: letmein
  parametersFrom:
    secretRef:
      name: mySecret
    secretKeyRef:
      name: myOtherSecret
      key: key1
```

duglin replied:
yes I would much prefer something like this. Moving the union up a level allows for the simple case to remain simple - meaning the case of just passing in a set of name/value pairs would be done by specifying them as children of "parameters" w/o an "inline" wrapper.

As we discussed off-line, what do people think of something like this (keep in mind this shows/covers all possible scenarios):   https://gist.github.com/duglin/1422e0eafc7eca22157af96a3f4a84d0

arschles replied:
@nilebox @duglin what did we settle on for this PR?

duglin commented on pkg/controller/controller_binding.go:
```diff
@@ -166,7 +166,7 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) error {
 		var parameters map[string]interface{}
 		if binding.Spec.Parameters != nil {
-			parameters, err = unmarshalParameters(binding.Spec.Parameters.Raw)
+			parameters, err = unmarshalRawParameters(binding.Spec.Parameters.Raw)
```
Why doesn't Binding use the same Parameters objects as Instance?  I see it defined as `Parameters *runtime.RawExtension` in the BindingSpec.

nilebox replied:
@duglin I'm planning to do this as a separate PR once we merge this one. One step at a time :)

duglin commented on pkg/controller/controller_instance_test.go:
```diff
@@ -215,7 +215,63 @@ func TestReconcileInstanceNonExistentServicePlan(t *testing.T) {
+	// Secret with invalid JSON
+	invalidSecret := &v1.Secret{
+		Data: map[string][]byte{
+			testSecretKey: []byte("invalid json"),
```
This is what I was talking about above. As a user I'm not sure it makes sense to ask me to encode my data as json when I should be able to just put the data in the secret as normal name/value pairs.

On today's call it might help if we could talk about these parameter properties at a higher level - meaning show what the UX needs are and would look like - then agree on those flows first.

Off the top of my head I think I've heard, or can think, of the following cases:
- user creates a secret to hold private data needed as params to instance/binding.create() call
- user has non-secret parameters that are needed to be passed to xxx.create() - might be alone or along with private creds from previous bullet
- user wants to use the binding/creds from one service binding as input params to another xxx.create() call
- user may need to add additional parameters (beyond the previous bullet's) to the xxx.create() call
- the keys used in parameters may use chars that are not acceptable by configMaps or secrets, so the user may need to store all/some of the param data as raw json.

Does this sound right? Are there others?

spadgett replied:
Can we determine if a value is a string, boolean, or number if it isn't encoded? Is secret value `true` the boolean `true` or string `"true"`?

duglin replied:
I'm having flashbacks to my days of working on Docker build where we tried to automagically determine if the CMD was a string/shell-cmd vs json array and we couldn't tell because things that started with '[' could be either.    :-)   The real trick was determining when to generate an error because you can't tell if a malformed json object was bad json vs just a string that happens to almost be valid json.    I think being explicit (and not trying to guess) is the safest.

nilebox replied:
@duglin yeah, I see what you mean, and I tend to agree that we should treat values as just arbitrary strings.

> Can we determine if a value is a string, boolean, or number if it isn't encoded? Is secret value true the boolean true or string "true"?

@spadgett  I don't think we can and should. Only if we apply a priority list of "duck typing" rules (and I don't like this idea, just giving a **bad** example of supporting this):
- if the string contains valid JSON, treat it as a `JSON tree`
- if the string is equal to "true" or "false", it is `boolean`
- if the string contains only digits, it is `int`
- ...
- if there is no match then, well... string is a string

Again, I think it is a bad idea. What if the user wanted to pass "true" as a string?
I would prefer to treat the value always the same way. And to make it easier for the users to use, I propose to treat multi-key secrets as Map<String, String> (without any magic duck typing). Any other use case is solved by single-key JSON blob, where you can put whatever you want, and we will pass it "as is".

duglin replied:
+1 to map<string,string> with an escape hatch of a json blob for the other/"as is" case.

spadgett replied:
> +1 to map<string,string> with an escape hatch of a json blob for the other/"as is" case.

Works for me.

I was arguing against duck typing and for JSON encoding in case I wasn't clear. But I'm also OK treating values always as strings for multi-key secrets.

nilebox replied:
Ok, going to implement this today.
Hopefully this will be the last change for this PR (and all other improvements are to be addressed through separate PRs)

nilebox replied:
Done

nilebox commented:
/retest

nilebox commented:
Substituted with revised proposal https://github.com/kubernetes-incubator/service-catalog/pull/1079
